internal/driver/mobile/android.c

```diff
@@ -390,17 +390,18 @@ char* listContentURI(uintptr_t jni_env, uintptr_t ctx, char* uriCstr) {
 	jclass cursorClass = (*env)->GetObjectClass(env, cursor);
 	jmethodID next = find_method(env, cursorClass, "moveToNext", "()Z");
 	jmethodID get = find_method(env, cursorClass, "getString", "(I)Ljava/lang/String;");
+	jmethodID getChildURI = find_static_method(env, contractClass, "buildDocumentUriUsingTree", "(Landroid/net/Uri;Ljava/lang/String;)Landroid/net/Uri;");
 
 	char *ret = NULL;
 	int len = 0;
 	while (((jboolean)(*env)->CallBooleanMethod(env, cursor, next)) == JNI_TRUE) {
-		jstring name = (jstring)(*env)->CallObjectMethod(env, cursor, get, 0);
-		jobject childUri = (jobject)(*env)->CallStaticObjectMethod(env, contractClass, getChild, uri, name);
+		jstring childDocId = (jstring)(*env)->CallObjectMethod(env, cursor, get, 0);
+		jobject childUri = (jobject)(*env)->CallStaticObjectMethod(env, contractClass, getChildURI, uri, childDocId);
 		jclass uriClass = (*env)->GetObjectClass(env, childUri);
 		jmethodID toString = (*env)->GetMethodID(env, uriClass, "toString", "()Ljava/lang/String;");
-		jobject s = (*env)->CallObjectMethod(env, childUri, toString);
+		jstring s = (jstring)(*env)->CallObjectMethod(env, childUri, toString);
 
-		char *uid = getString(jni_env, ctx, name);
+		char *uid = getString(jni_env, ctx, s);
```

Comment:
Look at it more carefully I wonder if it was a typo - with just this 1 line change does it not return the expected data?
(I like the move to jstring above, but I don't know if the additional lookup method is needed)

Reply:
Absolutely, changing name into s in line 404 partially solve the original problem:

Here lies another bug: the returned URIs are "children" URIs (because built using [buildChildDocumentsUriUsingTree](https://developer.android.com/reference/android/provider/DocumentsContract#buildChildDocumentsUriUsingTree(android.net.Uri,%20java.lang.String)) instead of [buildDocumentUriUsingTree](https://developer.android.com/reference/android/provider/DocumentsContract#buildDocumentUriUsingTree(android.net.Uri,%20java.lang.String))). They are valid and accessible, but lost the original file name and extension.

An example would make this more concrete:
If the parent directory URI is:
> content://com.android.externalstorage.documents/tree/primary:Android/media/test

When fixing only the typo, the child URI becomes:
> content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%2Ftest/document/primary%3AAndroid%2Fmedia%2Ftest%2Ftest.md/children

Notice the suffix "/children".
With this URI, calling URI.Name() returns "children" and URI.Extension() does not return the original extension (".md"). If the directory contains several files, they are all named "children"... This is not convenient.

When using buildDocumentUriUsingTree, the child URI is:
> content://com.android.externalstorage.documents/tree/primary%3AAndroid%2Fmedia%2Ftest/document/primary%3AAndroid%2Fmedia%2Ftest%2Ftest.md

Now each child has a unique name and the expected extension.

PS: I wish those URIs were simpler and their path not encoded twice.

Reply:
Good catch, thanks for the clarification